### PR TITLE
feat: adding `ConsensusHost` interface for custom self client/consensus state validation

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

Co-authored-by: chatton <github.qpeyb@simplelogin.fr>

Adds the `ConsensusHost` interface type to 02-client as well as implementation for `07-tendermint` and `08-wasm`.

closes: #5315

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

## Summary by CodeRabbit

- **New Features**
	- Improved blockchain interoperability by introducing a `WasmConsensusHost` struct for wasm-wrapped consensus in the IBC package.
- **Tests**
	- Ensured reliability and robustness through extensive testing of the new wasm consensus mechanism in diverse scenarios.

### CODE REVIEW

File: modules/core/keeper/keeper.go [hunk 1]
```diff
@@ -60,13 +60,13 @@ func NewKeeper(
 	}
 
 	clientKeeper := clientkeeper.NewKeeper(cdc, key, paramSpace, stakingKeeper, upgradeKeeper)
```
Comment (damiannolan) [inline modules/core/keeper/keeper.go:62]: Ideally, I think we should be replacing the `stakingKeeper` arg to core ibc's `NewKeeper` with a `types.ConsensusHost` and have users pipe in whatever consensus host they are using explicitly in app.go. 

E.g. `ibctm.NewConsensusHost(app.StakingKeeper)` in app.go as an arg to `ibc.NewKeeper(...)`

thoughts?

File: modules/core/keeper/keeper.go [hunk 2]
```diff
@@ -28,7 +28,7 @@ type Keeper struct {
 
 	cdc codec.BinaryCodec
 
-	ClientKeeper     clientkeeper.Keeper
+	ClientKeeper     *clientkeeper.Keeper
 	ConnectionKeeper connectionkeeper.Keeper
 	ChannelKeeper    channelkeeper.Keeper
 	PortKeeper       *portkeeper.Keeper
```
Comment (damiannolan) [inline modules/core/keeper/keeper.go:34]: @chatton and I have been discussing the mismatch of pointer and value types used here for a while, I think others are also aware of it - this was an issue with the `PortKeeper` below and caused problems with the router in the past.

@chatton volunteered to open an issue about this for discussion. I would be in favour of having consistency here, which would minimise the amount of debugging when things aren't quite lining up with what we expect to happen.

I think its probably best to not be dereferencing keepers when passing args all over the place - i.e. `*k.ClientKeeper` and such

File: modules/light-clients/07-tendermint/consensus_host.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	commitmenttypes "github.com/cosmos/ibc-go/v8/modules/core/23-commitment/types"
+	ibcerrors "github.com/cosmos/ibc-go/v8/modules/core/errors"
+	"github.com/cosmos/ibc-go/v8/modules/core/exported"
+)
+
+var _ clienttypes.ConsensusHost = (*ConsensusHost)(nil)
+
+// ConsensusHost implements the 02-client clienttypes.ConsensusHost interface
+type ConsensusHost struct {
+	stakingKeeper clienttypes.StakingKeeper
```
Comment (damiannolan) [inline modules/light-clients/07-tendermint/consensus_host.go:23]: todo: this expected keeper interface for `StakingKeeper` should be moved to 07-tendermint as well

Comment (damiannolan) [reply modules/light-clients/07-tendermint/consensus_host.go]: redefined in 07-tendermint, should be removed from clienttypes if and when the `NewKeeper` arg is changed.

see: https://github.com/cosmos/ibc-go/pull/6055#discussion_r1538903188

File: modules/core/keeper/keeper.go (hunk 2 above)
Comment (damiannolan) [reply modules/core/keeper/keeper.go:34]: #6058

File: modules/light-clients/08-wasm/types/consensus_host.go
Comment (damiannolan) [inline modules/light-clients/08-wasm/types/consensus_host.go:1]: needs tests

**Review by crodriguezvega — COMMENTED** [review]
Thanks for pulling out these changes from the feat branch, @damiannolan and @chatton. I will review more thoroughly later today, but the changes are looking good to me.

There are theoretically several API breaking changes, but I wonder if we need to document all of them, since I am not sure the changes are in functions that users will use? Like for example the change in `BeginBlocker` or in the client keeper `NewMigrator` function... What do you guys think? And do you think it will be manageable to revert these API breaking changes when back porting to v8.2?

I can also start working on adding docs for this and open a separate PR for it. 

We should also add changelog for this PR. :)

**Review by AdityaSripal — APPROVED** [review]
Approved, will the work on routing and client modules remove the need for a separate 08-wasm consensus host?

Ideally a tendermint chain can support counterparty clients being implemented in go or wasm

**Comment by damiannolan** [comment]
> Approved, will the work on routing and client modules remove the need for a separate 08-wasm consensus host?
> 
> Ideally a tendermint chain can support counterparty clients being implemented in go or wasm

Pasting what I wrote in dms for documenting purposes:

I’ll try to refresh my head a bit around the wasm client state wrapper a bit more, but looking briefly - its because the wrapped type (envelope) still contains the crucial part of info we need for instantiating the contract - the checksum. And currently, even with the changes we’ve made for client routing, we don’t have a way to plumb through this “secondary level routing info” through the client APIs.
However, I think the wrapper is only technically needed for this routing in the beginning with `Initialize` / `instatiateContract`.
The relayer can probably always pass the envelope on `CreateClient` but the contract could then technically store its own client state as the unwrapped client. (I think it would still need to be marshalled to an `Any` tho, in order to conform to the rest of the ibc code - e.g. I think this is important for `UpgradeClient`)

At the moment, contracts are actually prevented from storing whatever they want due to this function (so this would need to be modified): https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/08-wasm/types/vm.go#L235-L263

File: modules/light-clients/07-tendermint/consensus_host_test.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+func (suite *TendermintTestSuite) TestGetSelfConsensusState() {
+	var height clienttypes.Height
+
+	cases := []struct {
+		name     string
+		malleate func()
+		expError error
+	}{
+		{
```
[the remark this answers is not in the document]
Comment (damiannolan) [reply modules/light-clients/07-tendermint/consensus_host_test.go:24]: done!

File: modules/light-clients/08-wasm/types/consensus_host_test.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+			malleate: func() {},
+			expError: nil,
+		},
+		{
+			name: "failure: delegate error",
+			malleate: func() {
+				consensusHost.(*mock.ConsensusHost).GetSelfConsensusStateFn = func(ctx sdk.Context, height exported.Height) (exported.ConsensusState, error) {
+					return nil, mock.MockApplicationCallbackError
+				}
+			},
```
Comment (charleenfei) [inline modules/light-clients/08-wasm/types/consensus_host_test.go:35]: could do the cdc := nil, failure to marshal consensusState cases here too

File: modules/light-clients/07-tendermint/consensus_host_test.go [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+			name: "zero height",
+			malleate: func() {
+				height = clienttypes.ZeroHeight()
```
Comment (charleenfei) [inline modules/light-clients/07-tendermint/consensus_host_test.go:27]: this line is probably not neecessary

**Review by charleenfei — APPROVED** [review]
super nice work @chatton @damiannolan <3

**Comment by damiannolan** [comment]
Let's add something in migration doc in a follow up. cc @chatton

**Pull request merged**